### PR TITLE
ALB004 - Alignment container

### PR DESCRIPTION
**Pull request by JTurbakiewicz**

### CODE REVIEW

**Review by pjanck — CHANGES_REQUESTED** [review]

File: features/ALB004_Alignment-in-spatial-structure.feature [hunk 1]
```diff
@@ -0,0 +1,11 @@
+@implementer-agreement
+@ALB
+Feature: ALB004 - Alignment in spatial structure
+The rule verifies, that each IfcAlignment is contained in an IfcProject, but not contained in spatial structure.
```
[the remark this answers is not in the document]
Comment (JTurbakiewicz) [reply features/ALB004_Alignment-in-spatial-structure.feature:4]: Agreed. I am not yet 100% sure what the intended behavior established on the implementors forum is. 
@evandroAlfieri, could you please give a helping hand on this one?

File: features/resources/spatial_CompositionTable.csv [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+IfcRailway,IfcSite
+IfcRailway,IfcRailway
+IfcRoadPart,IfcRoad
+IfcRoad,IfcProject
+IfcRoad,IfcSite
+IfcRoad,IfcRoad
+IfcSite,IfcProject
+IfcSite,IfcSite
+IfcSpace,IfcProject
+IfcSpace,IfcSite
```
[the remark this answers is not in the document]
Comment (JTurbakiewicz) [reply features/resources/spatial_CompositionTable.csv:33]: It comes from: https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/concepts/Object_Composition/Aggregation/Spatial_Composition/content.html#General-Usage

File: features/steps/thens/relations.py [hunk 3]
```diff
@@ -36,10 +36,54 @@ def step_impl(context, entity, condition, directness, other_entity):
...
+            aggregated_table[applicable_entity].append(tbl_relationship_object)
+        else:
+            aggregated_table[applicable_entity] = [tbl_relationship_object]
+    errors = []
+
+    if context.instances and getattr(context, 'applicable', True):
+        for ent in context.model.by_type(entity):
+            for applicable_entity in aggregated_table.keys(): # check which applicable entity the currently processed entity is (inheritance), e.g IfcRailway -> IfcFacility
+                if ent.is_a(applicable_entity):
+                    break # TODO - > is the hierarchy important?
```
[the remark this answers is not in the document]
Comment (JTurbakiewicz) [reply features/steps/thens/relations.py:68]: Agreed. More discussion here: https://github.com/buildingSMART/ifc-gherkin-rules/pull/65#discussion_r1193218299

File: features/ALB004_Alignment-in-spatial-structure.feature [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+  Scenario: Agreement on each IfcAlignment being contained in an IfcProject and not in spatial structure
+
+      Given A file with Schema Identifier "IFC4X3_TC1" or "IFC4X3_ADD1" or "IFC4X3"
+      And An IfcAlignment
+      Then Each IfcAlignment must be aggregated to IfcProject as per spatial_CompositionTable.csv
+      And Each IfcAlignment must not be directly or indirectly contained in IfcSpatialElement
```
[the remark this answers is not in the document]
Comment (JTurbakiewicz) [reply features/ALB004_Alignment-in-spatial-structure.feature:11]: Ad.1. You're right. I thought it got mapped to IfcSpace, which connects to IfcProject in the .csv file. But it wasn't so, just the loop ended on IfcSpace, as it was the last element in the .csv. I'll fix that.

[the remark this answers is not in the document]
Comment (JTurbakiewicz) [reply features/ALB004_Alignment-in-spatial-structure.feature:11]: Ad 2. The intention was:

Directly: IfcAlignment contained in IfcSpatialElement
Indirectly: IfcAlginment contained in anything but the IfcSpatialElement, which then is contained in IfcSpatialElement. However many levels there would be.

File: features/ALB004_Alignment-in-spatial-structure.feature (hunk 1 above)
[the remark this answers is not in the document]
Comment (evandroAlfieri) [reply features/ALB004_Alignment-in-spatial-structure.feature:4]: Hi all, the promised update from the IF:
1. All instances of _IfcAlignment_ must be related to _IfcProject_ using the _IfcRelAggregates_ relationship - either directly or indirectly. The indirect case is when a child alignment is aggregated to a parent alignment. In this case, only the parent alignment shall be related to the project.
2. Instances of _IfcAlignment_ must not be related to spatial entities using the _IfcRelContainedInSpatialStructure_ relationship.

Can you try to adapt the rule accordingly?

File: features/ALB004_Alignment-in-spatial-structure.feature (hunk 4 above)
[the remark this answers is not in the document]
Comment (JTurbakiewicz) [reply features/ALB004_Alignment-in-spatial-structure.feature]: Changed the logic completely.

File: features/ALB004_Alignment-in-spatial-structure.feature
Comment (pjanck) [inline features/ALB004_Alignment-in-spatial-structure.feature:1]: Considering

> The indirect case is when a child alignment is aggregated to a parent alignment.

and 

> Then Each IfcAlignment must be aggregated to IfcProject directly or indirectly

together with the code provided in the `features/steps/thens/relations.py`, I wonder what a hypothetical situation like the one below would produce:

> IfcProject --> IfcRelAggregates --> IfcSite --> IfcRelAggregates --> IfcAlignment 
> (project is aggregating a site which in turn aggregates an alignment)

Is this passing or failing? (It should fail in my opinion.)

**Pull request merged**